Source/Particles/Collision/BinaryCollision/VirtualPhotonCreation.H

```diff
@@ -0,0 +1,147 @@
+/* Copyright 2021 Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#ifndef WARPX_VIRTUAL_PHOTONS_H_
+#define WARPX_VIRTUAL_PHOTONS_H_
+
+#include "Particles/MultiParticleContainer.H"
+#include "Particles/WarpXParticleContainer.H"
+#include "Particles/PhysicalParticleContainer.H"
+#include "Particles/ParticleCreation/SmartCopy.H"
+
+#include "WarpX.H"
+#include "Utils/TextMsg.H"
+
+
+#include <AMReX_INT.H>
+#include <AMReX_REAL.H>
+#include <AMReX_Particle.H>
+
+
+namespace collision::binarycollision::virtualphotons{
+
+using namespace amrex::literals;
+using SoaData_type = typename WarpXParticleContainer::ParticleTileType::ParticleTileDataType;
+
+void GenerateVirtualPhotons (MultiParticleContainer* mypc){
+
+    WARPX_PROFILE("collision::binarycollision::virtualphotons::GenerateVirtualPhotons()");
+
+    // Enable tiling
+    amrex::MFItInfo info;
+    if (amrex::Gpu::notInLaunchRegion()) { info.EnableTiling(WarpXParticleContainer::tile_size); }
+
+    // TO DO: particle species should be automatically selected
+    auto& leptons  = mypc->GetParticleContainerFromName("beam");
+    auto& vphotons = mypc->GetParticleContainerFromName("virtual_photons");
+
+    int const nlevs = std::max(0, leptons.finestLevel()+1);
+    for (int lev = 0; lev < nlevs; ++lev) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+        for (amrex::MFIter mfi = leptons.MakeMFIter(lev, info); mfi.isValid(); ++mfi)
+        {
+            // Notation: _vp means virtual photon
+
+            // primary particles (leptons) in the current tile
+            ParticleTileType& ptile = leptons.ParticlesAt(lev, mfi);
+            const auto soa = ptile.getParticleTileData();
+
+            // number of primary particles in the current tile
+            amrex::Long const num = ptile.numParticles();
+
+            // vector that will contain the number of virtual photons for each primary particle
+            amrex::Gpu::DeviceVector<amrex::Long> num_vp(num, 0);
+            auto* num_vp_data = num_vp.dataPtr();
+
+            // First pass: compute the number of virtual photons for each primary particle
+            // and fill the corresponding vector
+            amrex::ParallelFor(num,
+             [=] AMREX_GPU_DEVICE (amrex::Long i) noexcept
+            {
+                amrex::ParticleReal ux = soa.m_rdata[PIdx::ux][i];
+                amrex::ParticleReal uy = soa.m_rdata[PIdx::uy][i];
+                amrex::ParticleReal uz = soa.m_rdata[PIdx::uz][i];
+
+                // same as Xsuite and guinea-pig
+                amrex::ParticleReal gamma = std::sqrt( 1.0_rt + ux*ux + uy*uy + uz*uz);
```

Comment:
```suggestion
                amrex::ParticleReal gamma = std::sqrt( 1.0_rt + (ux*ux + uy*uy + uz*uz)*invclightsq );
```
This is admittedly confusing, but WarpX stores `u` in units of `m/s` (i.e. `u` is `gamma*beta*c`), even though the user enters `u` as dimensionless. See for instance:
https://github.com/BLAST-WarpX/warpx/blob/development/Source/Particles/Pusher/UpdateMomentumVay.H#L34

Reply:
Just saw this, I did realize it eventually